### PR TITLE
RPS clean up and improvements

### PR DESCRIPTION
**Pull request by lijing0010**
## Description

    - Removed tmvp constraint for minigop8 under minigop16 config
    - Expand the RPS tables to allow for the implementation of other
    prediction structures

## Related issue
#207

### CODE REVIEW

**Comment by lijing0010** [comment]
added more comments

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 1]
```diff
@@ -1413,9 +1416,44 @@ void set_all_ref_frame_type(SequenceControlSet *sequence_control_set_ptr, Pictur
...
     }
 
+    int tmp = av1_rps->ref_dpb_index[ALT];
+    av1_rps->ref_dpb_index[ALT] = av1_rps->ref_dpb_index[ALT2];
+    av1_rps->ref_dpb_index[ALT2] = tmp;
+
+    uint64_t tmp1 = av1_rps->ref_poc_array[ALT];
+    av1_rps->ref_poc_array[ALT] = av1_rps->ref_poc_array[ALT2];
+    av1_rps->ref_poc_array[ALT2] = tmp1;
+    return;
```
[the remark this answers is not in the document]
Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: I planned to invert it at the beginning, but decided to leave it as it is to align with original.
OK, I'll invert ALT/ALT2

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 2]
```diff
@@ -1458,16 +1497,405 @@ void  Av1GenerateRpsInfo(
...
+        av1_rps->ref_poc_array[ALT] = get_ref_poc(context_ptr, picture_control_set_ptr->picture_number, flat_pred_struct[gop_i].ref_list1[1]);
+        av1_rps->ref_poc_array[ALT2] = av1_rps->ref_poc_array[BWD];
+
+        amend_rps(pred_position_ptr, av1_rps);
+        av1_rps->refresh_frame_mask = 1 << context_ptr->lay0_toggle;
+
+        if (picture_control_set_ptr->pred_struct_ptr->pred_type == EB_PRED_LOW_DELAY_P || picture_control_set_ptr->is_overlay) {
+            //P frames
+            av1_rps->ref_dpb_index[4] = av1_rps->ref_dpb_index[5] = av1_rps->ref_dpb_index[6] = av1_rps->ref_dpb_index[0];
+            av1_rps->ref_poc_array[4] = av1_rps->ref_poc_array[5] = av1_rps->ref_poc_array[6] = av1_rps->ref_poc_array[0];
```
[the remark this answers is not in the document]
Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: sure

**Comment by lijing0010** [comment]
> we have restrictions in the code for some features for some gop structres like picture_control_set_ptr->frm_hdr.use_ref_frame_mvs. this is for example Temporal MVP that depends on how the GOP is implemented. can you please make that those features are supported by the new added GOP structures

@chkngit ,why limit the tmvp for trailing frames in minigop16? any specific reasons?

**Comment by chkngit** [comment]
why limit the tmvp for trailing frames in minigop16? any specific reasons?

there were a  mismatch between RPS and the references uses in the encoder for the case of MiniGop8 in a mingop16 config. so TMVP will not work
so turning off TMVP only on that mingop8 was a workaorund to this problem.

**Comment by lijing0010** [comment]
@chkngit , I refined and rebased the code.
For the concern you raised about the tmvp, I kept the current logic for now(disable tmvp for minigop8 in minigop16 config. Although I think it should be fine)

**Comment by chkngit** [comment]
if you think there is no need to keep the restriction, please remove it. it will bring some bd rate gain!

**Comment by lijing0010** [comment]
@chkngit ,OK, will start another round of test, and submit the PR when it's done

**Comment by lijing0010** [comment]
@chkngit , please have a review, just confirmed that tmvp constraint can be removed

**Comment by hassount** [comment]
@lijing0010 
Restarted the Travis test. Please address @fouedb request for comments and rebase / squash the commits into the minimum amount of commits if not 1

**Comment by hassount** [comment]
@joelsole @fouedb 
Any thoughts on this PR?

**Comment by lijing0010** [comment]
@hassount , just did a quick check. tmvp with other encmode(1 to 8) are also fixed by this PR.
You can do a test on your side to verify it

File: Source/Lib/Common/Codec/EbPredictionStructure.c [hunk 3]
```diff
@@ -1040,7 +1040,9 @@ static EbErrorType PredictionStructureCtor(
             }
 
             // Set Leading Picture's Reference List 0 Count
-            predictionStructurePtr->pred_struct_entry_ptr_array[entryIndex]->ref_list0.reference_list_count = refIndex;
+            // av1 supports 3 backward, limit the number of refs for low_delay_b case
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbPredictionStructure.c]: Please make the comment more clear and precise.

Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPredictionStructure.c]: done

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 4]
```diff
@@ -3574,11 +3644,20 @@ void* picture_decision_kernel(void *input_ptr)
                                     if (!(*fgn_random_seed_ptr))     // Random seed should not be zero
                                         *fgn_random_seed_ptr += 7391;
                                 }
+                                uint32_t pic_index = 0;
+                                if (sequence_control_set_ptr->static_config.pred_structure == EB_PRED_RANDOM_ACCESS) {
+                                    pic_index = pictureIndex - context_ptr->mini_gop_start_index[mini_gop_index];
+                                } else {
+                                    // For low delay P or low delay B case, get the the picture_index by mini_gop size
+                                    pic_index = (picture_control_set_ptr->picture_number == 0) ? 0 :
+                                        (uint32_t)(((picture_control_set_ptr->picture_number - 1) % (sequence_control_set_ptr->static_config.intra_period_length+1)) % picture_control_set_ptr->pred_struct_ptr->pred_struct_period);
```
Comment (jerrymiao-tencent) [inline Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: may trigger issues with division by 0,  if intra_period is equal to -1.

**Comment by hassount** [comment]
@lijing0010 
Would you be able to provide some BDR numbers for 1L - 4L as compared to libaom even 1 pass to know where we are at this point? even the 360p and 720p clips

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c (hunk 4 above)
Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: thanks for the report, I'll fix it

**Comment by lijing0010** [comment]
> @lijing0010
> Would you be able to provide some BDR numbers for 1L - 4L as compared to libaom even 1 pass to know where we are at this point? even the 360p and 720p clips

I'll ask WangJing to help to run some tests on it

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c (hunk 4 above)
Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: done

**Comment by lijing0010** [comment]
@hassount , I did a quick test for pred-struct 1 and low latency in aom(both 1 pass) on 7 360P bitstreams. The quality gap is big, about 45%. Needs further tuning on the quality
The command for SVT is:
SvtAv1EncApp -enc-mode 0  -w 640  -h 360  -i /home/jingli/YUV/blue_sky_360p_60f.yuv  -q 20  -fps-num 25000 -fps-denom 1000 -b bitstreams/low_delay_test_M0_blue_sky_360p_60f_Q20.bin   -n 60  -intra-period 119  -lad 119  -hdr 0 -pred-struct 1

The command for aom is:
aomenc --width=640  --height=360  --cq-level=20  --fps=25000/1000 -o bitstreams/aom_test_M0_blue_sky_360p_60f_Q20.bin   /home/jingli/YUV/blue_sky_360p_60f.yuv --kf-max-dist=120  --codec=av1 --cpu-used=0 --threads=1  --end-usage=q --lag-in-frames=0 --psnr

**Comment by hassount** [comment]
> @hassount , I did a quick test for pred-struct 1 and low latency in aom(both 1 pass) on 7 360P bitstreams. The quality gap is big. Needs further tuning on the quality
> The command for SVT is:
> SvtAv1EncApp -enc-mode 0 -w 640 -h 360 -i /home/jingli/YUV/blue_sky_360p_60f.yuv -q 20 -fps-num 25000 -fps-denom 1000 -b bitstreams/low_delay_test_M0_blue_sky_360p_60f_Q20.bin -n 60 -intra-period 119 -lad 119 -hdr 0 -pred-struct 1
> 
> The command for aom is:
> aomenc --width=640 --height=360 --cq-level=20 --fps=25000/1000 -o bitstreams/aom_test_M0_blue_sky_360p_60f_Q20.bin /home/jingli/YUV/blue_sky_360p_60f.yuv --kf-max-dist=120 --codec=av1 --cpu-used=0 --threads=1 --end-usage=q --lag-in-frames=0 --psnr

Thank you Li Jing for running the test, let's make sure that we first align the QP / pred structure between them to get an apples-to-apples comparison. If the difference is still large, that means we'll have to debug and we should shut down the pres-struct option until such is done.

**Comment by lijing0010** [comment]
@hassount , you are right. 
This PR has just resolved a few RPS related constraint, haven't fine tuned it yet. 
It uses much fewer reference frames than libaom right now, and I will align the pred-struct to libAOM and do an apple to apple comparison.

Some part of the PR can remove the constraint on tmvp and can help to improve the BDrate for random access. Maybe I can separate it to another patch?

File: Source/Lib/Encoder/Codec/EbEncHandle.c [hunk 5]
```diff
@@ -2042,7 +2042,7 @@ void CopyApiFromApp(
     sequence_control_set_ptr->general_interlaced_source_flag = 0;
 
     // SB Definitions
-    sequence_control_set_ptr->static_config.pred_structure = 2; // Hardcoded(Cleanup)
+    sequence_control_set_ptr->static_config.pred_structure = pComponentParameterStructure->pred_structure;
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbEncHandle.c]: this should not be done yet, until we make the underlying infrastructure work properly

**Comment by hassount** [comment]
@lijing0010 can you please address @andrey-norkin comments so that we merge this ?

File: Source/Lib/Encoder/Codec/EbEncHandle.c (hunk 5 above)
Comment (lijing0010) [reply Source/Lib/Encoder/Codec/EbEncHandle.c]: I disabled it in VerifySettings, but I can also disable it here

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 6]
```diff
@@ -1436,8 +1438,94 @@ void set_all_ref_frame_type(SequenceControlSet *sequence_control_set_ptr, Pictur
...
+        av1_rps->ref_poc_array[ALT] = av1_rps->ref_poc_array[BWD];
+    }
+    if (pred_position_ptr->ref_list1.reference_list_count < 2) {
+        av1_rps->ref_dpb_index[ALT2] = av1_rps->ref_dpb_index[BWD];
+        av1_rps->ref_poc_array[ALT2] = av1_rps->ref_poc_array[BWD];
+    }
+}
+
+// Handle the show mechanism for 1)Low delay P, 2)Low delay B and 3)I frames of RA
+static EbBool handle_show_mechanism(
```
[the remark this answers is not in the document]
Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: yeah, make sense, set_frame_display_params looks better

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 7]
```diff
@@ -1436,8 +1438,94 @@ void set_all_ref_frame_type(SequenceControlSet *sequence_control_set_ptr, Pictur
...
+            if (context_ptr->mini_gop_length[mini_gop_index] < picture_control_set_ptr->pred_struct_ptr->pred_struct_period) {
+                //Scene Change that breaks the mini gop and switch to LDP (if I scene change happens to be aligned with a complete miniGop, then we do not break the pred structure)
+                frm_hdr->show_frame = EB_TRUE;
+                picture_control_set_ptr->has_show_existing = EB_FALSE;
+            } else {
+                frm_hdr->show_frame = EB_FALSE;
+                picture_control_set_ptr->has_show_existing = EB_FALSE;
+            }
+        } else {
+			if (context_ptr->mini_gop_length[0] != picture_control_set_ptr->pred_struct_ptr->pred_struct_period) {
```
[the remark this answers is not in the document]
Comment (lijing0010) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: done

**Comment by hassount** [comment]
@lijing0010 thank you for the changes and rebasing 
@andrey-norkin is this good to go from your side?

**Comment by hassount** [comment]
@lijing0010 
Can we get this rebased on top of the latest master so that it's ready to be merged by the code freeze ?

**Pull request merged**